src/wasm/wasm-validator.cpp

```diff
@@ -2767,13 +2767,21 @@ void FunctionValidator::visitFunction(Function* curr) {
   }
 }
 
-static bool checkSegmentOffset(Expression* curr, Address add, Address max) {
-  if (curr->is<GlobalGet>()) {
-    return true;
+static bool checkSegmentOffset(Expression* curr,
+                               Address add,
+                               Address max,
+                               FeatureSet features) {
+  if (!Properties::isValidInConstantExpression(curr, features)) {
+    return false;
   }
   auto* c = curr->dynCast<Const>();
   if (!c) {
-    return false;
+    // Unless the instruction is actually a const instruction, we don't
+    // currently try to evaluate it.
+    // TODO: Attempt to evaluate other expressions that might also be const
```

Comment:
```suggestion
    // TODO: Attempt to validate other expressions that might also be const
```

Reply:
Actually the validation is done already.. this part of the code is trying to evaluate to a absolute constant.  Does that make sense?

Reply:
To clarify, the validation of the instructions is done in isValidInConstantExpression.. this part is trying to evaluate the expression to get an absolute offset so that we can bounds check it.    This not always possible... for example if the expression is a global.get we currently just skip the bounds check and assume is not possible to evaluate.

Reply:
Oh, right... yes, ignore my comments then. lgtm as is.